File: Extra Credit/JavaScript/app/infrastructure/ActivityController.js
class ActivityController {
    constructor(mainActivityClass) {
        this.activityStack = new Array();
        // the mainActivity will always be on the stack
        Activity.controller = this;
        this.startActivity(mainActivityClass);
    }

    get currentActivity() {
        return this.activityStack[this.activityStack.length - 1];
    }

    startActivity(activityClass) {
        if (this.activityStack.length > 0) {
            this.currentActivity.pause();
        }
        this.activityStack.push(new activityClass());
        this.currentActivity.start();
    }

    terminateActivity(result) {
        if (this.activityStack.length !== 1) {
            this.currentActivity.terminate();
            this.activityStack.pop();
            this.currentActivity.resume(result);
        } else {
            throw "Cannot terminate main activity.";
        }
    }

    update() {
        this.currentActivity.update();
    }

    draw() {
        this.currentActivity.draw();
    }

    handleMousePressed(x, y) {
        this.currentActivity.handleMouseReleased(x, y);
    }

    handleMouseReleased(x, y) {
        this.currentActivity.handleMouseReleased(x, y);
    }

    handleMouseClicked(x, y) {
        this.currentActivity.handleMouseClicked(x, y);
    }

    handleResize() {
        this.currentActivity.handleResize();
    }

    handleKeyPressed(key, keyCode) {
        this.currentActivity.handleKeyPressed(key, keyCode);
    }

    handleKeyTyped(key, keyCode) {
        this.currentActivity.handleKeyTyped(key, keyCode);
    }
}

File: Extra Credit/Python/view/activitycontroller.py
class ActivityController:
    def __init__(self, term):
        self.term = term
        self.activity_stack = []

    @property
    def current_activity(self):
        return self.activity_stack[-1]

    def start_activity(self, activity):
        print(self.term.clear(), end="")
        self.activity_stack.append(activity(self.term, self))
        self.current_activity.start()

    def terminate_activity(self, res):
        cls = type(self.activity_stack.pop())
        if len(self.activity_stack) > 0:
            self.current_activity.resume(res, cls)
